File: packages/firebase_common/lib/src/firestore_documents/worker.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'worker.dart';

class ReadWorker {
  const ReadWorker({
    required this.workerId,
    required this.path,
    required this.displayName,
    required this.imageUrl,
    required this.introduction,
    required this.isHost,
    required this.createdAt,
    required this.updatedAt,
  });

  final String workerId;

  final String path;

  final String displayName;

  final String imageUrl;

  final String introduction;

  final bool isHost;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadWorker._fromJson(Map<String, dynamic> json) {
    return ReadWorker(
      workerId: json['workerId'] as String,
      path: json['path'] as String,
      displayName: json['displayName'] as String? ?? '',
      imageUrl: json['imageUrl'] as String? ?? '',
      introduction: json['introduction'] as String? ?? '',
      isHost: json['isHost'] as bool? ?? false,
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadWorker.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadWorker._fromJson(<String, dynamic>{
      ...data,
      'workerId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateWorker {
  const CreateWorker({
    required this.displayName,
    this.imageUrl = '',
    this.introduction = '',
    required this.isHost,
  });

  final String displayName;
  final String imageUrl;
  final String introduction;
  final bool isHost;

  Map<String, dynamic> toJson() {
    return {
      'displayName': displayName,
      'imageUrl': imageUrl,
      'introduction': introduction,
      'isHost': isHost,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateWorker {
  const UpdateWorker({
    this.displayName,
    this.imageUrl,
    this.introduction,
    this.isHost,
    this.createdAt,
  });

  final String? displayName;
  final String? imageUrl;
  final String? introduction;
  final bool? isHost;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (displayName != null) 'displayName': displayName,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (introduction != null) 'introduction': introduction,
      if (isHost != null) 'isHost': isHost,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteWorker {}

/// Provides a reference to the workers collection for reading.
final readWorkerCollectionReference =
    FirebaseFirestore.instance.collection('workers').withConverter<ReadWorker>(
          fromFirestore: (ds, _) => ReadWorker.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a worker document for reading.
DocumentReference<ReadWorker> readWorkerDocumentReference({
  required String workerId,
}) =>
    readWorkerCollectionReference.doc(workerId);

/// Provides a reference to the workers collection for creating.
final createWorkerCollectionReference = FirebaseFirestore.instance
    .collection('workers')
    .withConverter<CreateWorker>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a worker document for creating.
DocumentReference<CreateWorker> createWorkerDocumentReference({
  required String workerId,
}) =>
    createWorkerCollectionReference.doc(workerId);

/// Provides a reference to the workers collection for updating.
final updateWorkerCollectionReference = FirebaseFirestore.instance
    .collection('workers')
    .withConverter<UpdateWorker>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a worker document for updating.
DocumentReference<UpdateWorker> updateWorkerDocumentReference({
  required String workerId,
}) =>
    updateWorkerCollectionReference.doc(workerId);

/// Provides a reference to the workers collection for deleting.
final deleteWorkerCollectionReference = FirebaseFirestore.instance
    .collection('workers')
    .withConverter<DeleteWorker>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a worker document for deleting.
DocumentReference<DeleteWorker> deleteWorkerDocumentReference({
  required String workerId,
}) =>
    deleteWorkerCollectionReference.doc(workerId);

/// Manages queries against the workers collection.
class WorkerQuery {
  /// Fetches [ReadWorker] documents.
  Future<List<ReadWorker>> fetchDocuments({
    GetOptions? options,
    Query<ReadWorker>? Function(Query<ReadWorker> query)? queryBuilder,
    int Function(ReadWorker lhs, ReadWorker rhs)? compare,
  }) async {
    Query<ReadWorker> query = readWorkerCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [Worker] documents.
  Stream<List<ReadWorker>> subscribeDocuments({
    Query<ReadWorker>? Function(Query<ReadWorker> query)? queryBuilder,
    int Function(ReadWorker lhs, ReadWorker rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadWorker> query = readWorkerCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadWorker] document.
  Future<ReadWorker?> fetchDocument({
    required String workerId,
    GetOptions? options,
  }) async {
    final ds = await readWorkerDocumentReference(
      workerId: workerId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [Worker] document.
  Stream<ReadWorker?> subscribeDocument({
    required String workerId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readWorkerDocumentReference(
      workerId: workerId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [Worker] document.
  Future<DocumentReference<CreateWorker>> add({
    required CreateWorker createWorker,
  }) =>
      createWorkerCollectionReference.add(createWorker);

  /// Sets a [Worker] document.
  Future<void> set({
    required String workerId,
    required CreateWorker createWorker,
    SetOptions? options,
  }) =>
      createWorkerDocumentReference(
        workerId: workerId,
      ).set(createWorker, options);

  /// Updates a specific [Worker] document.
  Future<void> update({
    required String workerId,
    required UpdateWorker updateWorker,
  }) =>
      updateWorkerDocumentReference(
        workerId: workerId,
      ).update(updateWorker.toJson());

  /// Deletes a specific [Worker] document.
  Future<void> delete({
    required String workerId,
  }) =>
      deleteWorkerDocumentReference(
        workerId: workerId,
      ).delete();
}

File: functions/src/firestore-documents/worker.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadWorker {
    constructor({
        workerId,
        path,
        displayName,
        imageUrl,
        isHost,
        createdAt,
        updatedAt
    }: {
        workerId: string
        path: string
        displayName: string
        imageUrl: string
        isHost: boolean
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.workerId = workerId
        this.path = path
        this.displayName = displayName
        this.imageUrl = imageUrl
        this.isHost = isHost
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly workerId: string

    readonly path: string

    readonly displayName: string

    readonly imageUrl: string

    readonly isHost: boolean

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static fromJson(json: Record<string, unknown>): ReadWorker {
        return new ReadWorker({
            workerId: json[`workerId`] as string,
            path: json[`path`] as string,
            displayName: (json[`displayName`] as string | undefined) ?? ``,
            imageUrl: (json[`imageUrl`] as string | undefined) ?? ``,
            isHost: (json[`isHost`] as boolean | undefined) ?? false,
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadWorker {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadWorker.fromJson({
            ...cleanedData,
            workerId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateWorker {
    constructor({
        displayName,
        imageUrl,
        isHost
    }: {
        displayName: string
        imageUrl: string
        isHost: boolean
    }) {
        this.displayName = displayName
        this.imageUrl = imageUrl
        this.isHost = isHost
    }

    readonly displayName: string

    readonly imageUrl: string

    readonly isHost: boolean

    toJson(): Record<string, unknown> {
        return {
            displayName: this.displayName,
            imageUrl: this.imageUrl,
            isHost: this.isHost,
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateWorker {
    constructor({
        displayName,
        imageUrl,
        isHost,
        createdAt
    }: {
        displayName?: string
        imageUrl?: string
        isHost?: boolean
        createdAt?: Date
    }) {
        this.displayName = displayName
        this.imageUrl = imageUrl
        this.isHost = isHost
        this.createdAt = createdAt
    }

    readonly displayName?: string

    readonly imageUrl?: string

    readonly isHost?: boolean

    readonly createdAt?: Date

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.displayName != undefined) {
            json[`displayName`] = this.displayName
        }
        if (this.imageUrl != undefined) {
            json[`imageUrl`] = this.imageUrl
        }
        if (this.isHost != undefined) {
            json[`isHost`] = this.isHost
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteWorker {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the workers collection for reading.
 */
export const readWorkerCollectionReference = db
    .collection(`workers`)
    .withConverter<ReadWorker>({
        fromFirestore: (ds: DocumentSnapshot): ReadWorker => {
            return ReadWorker.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(`toFirestore is not implemented for ReadWorker`)
        }
    })

/**
 * Provides a reference to a worker document for reading.
 * @param workerId - The ID of the worker document to read.
 */
export const readWorkerDocumentReference = ({
    workerId
}: {
    workerId: string
}): DocumentReference<ReadWorker> => readWorkerCollectionReference.doc(workerId)

/**
 * Provides a reference to the workers collection for creating.
 */
export const createWorkerCollectionReference = db
    .collection(`workers`)
    .withConverter<CreateWorker>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for CreateWorker`)
        },
        toFirestore: (obj: CreateWorker): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a worker document for creating.
 * @param workerId - The ID of the worker document to read.
 */
export const createWorkerDocumentReference = ({
    workerId
}: {
    workerId: string
}): DocumentReference<CreateWorker> =>
    createWorkerCollectionReference.doc(workerId)

/**
 * Provides a reference to the workers collection for updating.
 */
export const updateWorkerCollectionReference = db
    .collection(`workers`)
    .withConverter<UpdateWorker>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for UpdateWorker`)
        },
        toFirestore: (obj: UpdateWorker): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a worker document for updating.
 * @param workerId - The ID of the worker document to read.
 */
export const updateWorkerDocumentReference = ({
    workerId
}: {
    workerId: string
}): DocumentReference<UpdateWorker> =>
    updateWorkerCollectionReference.doc(workerId)

/**
 * Provides a reference to the workers collection for deleting.
 */
export const deleteWorkerCollectionReference = db
    .collection(`workers`)
    .withConverter<DeleteWorker>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for DeleteWorker`)
        },
        toFirestore: (): DocumentData => {
            throw new Error(`toFirestore is not implemented for DeleteWorker`)
        }
    })

/**
 * Provides a reference to a worker document for deleting.
 * @param workerId - The ID of the worker document to read.
 */
export const deleteWorkerDocumentReference = ({
    workerId
}: {
    workerId: string
}): DocumentReference<DeleteWorker> =>
    deleteWorkerCollectionReference.doc(workerId)

/**
 * Manages queries against the workers collection.
 */
export class WorkerQuery {
    /**
     * Fetches worker documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (query: Query<ReadWorker>) => Query<ReadWorker>
        compare?: (lhs: ReadWorker, rhs: ReadWorker) => number
    }): Promise<ReadWorker[]> {
        let query: Query<ReadWorker> = readWorkerCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadWorker> = await query.get()
        let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadWorker>) =>
            qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific worker document.
     * @param workerId - The ID of the worker document to fetch.
     */
    async fetchDocument({
        workerId
    }: {
        workerId: string
    }): Promise<ReadWorker | undefined> {
        const ds = await readWorkerDocumentReference({
            workerId
        }).get()
        return ds.data()
    }

    /**
     * Adds a worker document.
     * @param createWorker - The worker details to add.
     */
    async add({
        createWorker
    }: {
        createWorker: CreateWorker
    }): Promise<DocumentReference<CreateWorker>> {
        return createWorkerCollectionReference.add(createWorker)
    }

    /**
     * Sets a worker document.
     * @param workerId - The ID of the worker document to set.
     * @param createWorker - The worker details to set.
     * @param options - Options for the set operation.
     */
    async set({
        workerId,
        createWorker,
        options
    }: {
        workerId: string
        createWorker: CreateWorker
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createWorkerDocumentReference({
                workerId
            }).set(createWorker)
        } else {
            return createWorkerDocumentReference({
                workerId
            }).set(createWorker, options)
        }
    }

    /**
     * Updates a specific worker document.
     * @param workerId - The ID of the worker document to update.
     * @param updateWorker - The details for updating the worker.
     */
    async update({
        workerId,
        updateWorker
    }: {
        workerId: string
        updateWorker: UpdateWorker
    }): Promise<WriteResult> {
        return updateWorkerDocumentReference({
            workerId
        }).update(updateWorker.toJson())
    }

    /**
     * Deletes a specific worker document.
     * @param workerId - The ID of the worker document to delete.
     */
    async delete({ workerId }: { workerId: string }): Promise<WriteResult> {
        return deleteWorkerDocumentReference({
            workerId
        }).delete()
    }
}
